Data.Rmd

```diff
@@ -0,0 +1,462 @@
+-   It throws distinct errors when the package does not exist vs. when the file does not exist within the package.
+-   During development, it works for interactive calls, calls from within the loaded package's namespace, and even for calls originating in dependencies.
+
+```{r error = TRUE}
+fs::path_package("extdata", package = "idonotexist")
+
+fs::path_package("extdata", "I_do_not_exist.csv", package = "readr")
+
+fs::path_package("extdata", "chickens.csv", package = "readr")
+```
+
+### `pkg_example()` path helpers
+
+We like to offer convenience functions that make example files easy to access.
+These are just user-friendly wrappers around `system.file()` or `fs::path_package()`, but can have added features, such as the ability to list the example files.
+Here's the definition and some usage of `readxl::read_example()`:
+
+```{r, eval = FALSE}
+readxl_example <- function(path = NULL) {
+  if (is.null(path)) {
+    dir(system.file("extdata", package = "readxl"))
+  } else {
+    system.file("extdata", path, package = "readxl", mustWork = TRUE)
+  }
+}
+```
+
+```{r}
+readxl::readxl_example()
+
+readxl::readxl_example("clippy.xlsx")
+```
+
+## Internal state
+
+Sometimes there's information that multiple functions from your package need to access that:
+
+-   Must be determined at runtime, not build time. It might even be dynamic.
+-   Doesn't make sense to pass in via a function argument. Often it's some obscure detail that a user shouldn't even know about.
+
+A great way to manage such data is to use an *environment*.[^data-1]
+This environment must be created a build time, but you can populate it with values at runtime and update those values over the course of an R session.
+This works because environments have reference semantics (whereas more pedestrian R objects, such as atomic vectors, lists, or data frames have value semantics).
+
+[^data-1]: If you don't know much about R environments and what makes them special, a great resource is the [Environments chapter](https://adv-r.hadley.nz/environments.html) of Advanced R.
+
+Consider a package that can store the user's favorite letters or numbers.
+You might start out with code like this in a file below `R/`:
+
+```{r eval = FALSE}
+favorite_letters <- letters[1:3]
+
+#' Report my favorite letters
+#' @export
+mfl <- function() {
+  favorite_letters
+}
+
+#' Change my favorite letters
+#' @export
+set_mfl <- function(l = letters[24:26]) {
+  favorite_letters <<- l
+  favorite_letters
+}
+```
+
+`favorite_letters` is initialized to ("a", "b", "c") when the package is built.
+The user can then inspect `favorite_letters` with `mfl()`, at which point they'll probably want to register *their* favorite letters with `set_mfl()`.
+Note that we've used the super assignment operator `<<-` in `set_mfl()` in the hope that this will reach up into the package environment and modify the internal data object `favorite_letters`.
+But a call to `set_mfl()` fails like so:
```

Comment:
This code works if you run it in a regular R session, but if you put it inside a package and run it you'll see:

Reply:
added